tests/sentry/workflow_engine/migrations/test_0105_deduplicate_workflows.py

```diff
@@ -0,0 +1,447 @@
+
+        # create action
+        if config["actions"]:
+            # associate each action to the condition group
+            for action in config["actions"]:
+                action.save()
+                self.create_data_condition_group_action(
+                    action=action,
+                    condition_group=action_filter_group,
+                )
+        elif config["mock_actions"]:
+            action = self.create_action()
+            self.create_data_condition_group_action(
+                action=action,
+                condition_group=action_filter_group,
+            )
+
+        workflow.save()
+        return workflow
+
+    def setup_initial_state(self) -> None:
+        self.orgs = []
+
+        # Isolate each test case to an organization
+        for i, workflow_config in enumerate(DUPLICATE_WORKFLOW_CONFIGS):
+            org = self.create_organization(name=str(i))
+
+            # Create a random number of duplicate workflows for each org
+            for _ in range(randint(2, 5)):
+                self.mock_workflow(org, **deepcopy(workflow_config))
+
+            for default_config in DEFAULT_WORKFLOW_CONFIGS:
+                self.mock_workflow(org, **deepcopy(default_config))
+
+            self.orgs.append(org)
+
+    def validate_org_workflows_deduped(self, duplicate_workflow_index: int):
+        org = Organization.objects.annotate(workflow_count=Count("workflow")).get(
+            name=str(duplicate_workflow_index)
+        )
+        expected_count = (
+            len(DEFAULT_WORKFLOW_CONFIGS)
+            if DUPLICATE_WORKFLOW_CONFIGS[duplicate_workflow_index] in DEFAULT_WORKFLOW_CONFIGS
+            else len(DEFAULT_WORKFLOW_CONFIGS) + 1
+        )
+        assert org.workflow_count == expected_count
+
+        workflow_ids = org.workflow_set.values_list("id", flat=True)
+
+        # Ensures that we have updated all the connections for workflows that have been deduplicated
+        count_connections = DetectorWorkflow.objects.filter(workflow_id__in=workflow_ids).count()
+        assert count_connections > expected_count
+
+        count_legacy = AlertRuleWorkflow.objects.filter(workflow_id__in=workflow_ids).count()
+        assert count_legacy > expected_count
+
+        # The legacy connections should match the new connections
+        assert count_connections == count_legacy
+
+    def test_deduplication__all_mocks(self) -> None:
+        index = 0  # mock everything
+        self.validate_org_workflows_deduped(index)
+
+    def test_deduplication__no_triggers(self) -> None:
+        index = 1  # workflow only has triggers
+        self.validate_org_workflows_deduped(index)
+
+    def test_deduplication__no_action_filters(self) -> None:
+        index = 2  # workflow has triggers + filters, no actions
+        self.validate_org_workflows_deduped(index)
+
+    def test_deduplication__only_action_filters(self) -> None:
+        index = 3  # workflow only has filters
```

Comment:
i'm not a huge fan of this test structure cause it's pretty magical, but i also can't get `parametrize` to work either (it's related to our base TestCase and how we munge the method calls). any recommendations on the structure? (the other problem is, we can't call the mock creator here because the migration has already happened when the test case is called 🤔)

Reply:
Can you not do something like this?

```
for i in range(1, 11): 
    self.validate_org_workflows_deduped(i)
```